src/storage-client/src/types/sources.rs

```diff
@@ -103,6 +105,55 @@ impl<S> IngestionDescription<S> {
     }
 }
 
+impl<S: Debug + Eq + PartialEq> IngestionDescription<S> {
+    /// Determines if `self` is compatible with another `IngestionDescription`, such that we can
+    /// support an `ALTER SOURCE`.
+    pub fn alter_compatible(
+        &self,
+        id: GlobalId,
+        other: &IngestionDescription<S>,
+    ) -> Result<(), StorageError> {
+        let IngestionDescription {
+            desc,
+            source_imports,
+            ingestion_metadata,
+            source_exports,
+            instance_id,
+            remap_collection_id,
+        } = self;
+
+        self.desc.alter_compatible(id, desc)?;
+
+        for incompatible in [
+            source_imports == &other.source_imports,
+            ingestion_metadata == &other.ingestion_metadata,
+            source_exports
+                .iter()
+                .merge_join_by(&other.source_exports, |(l_key, _), (r_key, _)| {
+                    l_key.cmp(r_key)
+                })
+                .all(|r| match r {
+                    Both((_, l_val), (_, r_val)) => l_val == r_val,
+                    _ => true,
+                }),
+            instance_id == &other.instance_id,
+            remap_collection_id == &other.remap_collection_id,
+        ] {
```

Comment:
nit: whenever this happens (multi-line for declaration) you can store the iteratee in a variable and use that in the for loop which results in the same  lines of code and a clearer for definition:

```rust

        let compatibility_checks = [
            source_imports == &other.source_imports,
            ingestion_metadata == &other.ingestion_metadata,
            source_exports
                .iter()
                .merge_join_by(&other.source_exports, |(l_key, _), (r_key, _)| {
                    l_key.cmp(r_key)
                })
                .all(|r| match r {
                    Both((_, l_val), (_, r_val)) => l_val == r_val,
                    _ => true,
                }),
            instance_id == &other.instance_id,
            remap_collection_id == &other.remap_collection_id,
        ];
        for incompatible in compatibility_checks {
```